common/network-types/src/session/socket/state.rs

```diff
@@ -0,0 +1,552 @@
+    },
+};
+
+pub struct SocketComponents<const C: usize> {
+    pub inspector: Option<FrameInspector>,
+    pub ctl_tx: UnboundedSender<SessionMessage<C>>,
+}
+
+/// Abstraction of the [SessionSocket] state.
+pub trait SocketState<const C: usize>: Clone + Send {
+    /// Gets ID of this Session.
+    fn session_id(&self) -> &str;
+
+    /// Starts the necessary processes inside the state.
+    /// Should be idempotent if called multiple times.
+    fn run(&mut self, components: SocketComponents<C>) -> Result<(), SessionError>;
+
+    /// Stops processes inside the state for the given direction.
+    fn stop(&mut self) -> Result<(), SessionError>;
+
+    /// Called when the Socket receives a new segment from Downstream.
+    /// When the error is returned, the incoming segment is not passed Upstream.
+    fn incoming_segment(&mut self, id: &SegmentId, segment_count: SeqNum) -> Result<(), SessionError>;
+
+    /// Called when [segment retransmission request](SegmentRequest) is received from Downstream.
+    fn incoming_retransmission_request(&mut self, request: SegmentRequest<C>) -> Result<(), SessionError>;
+
+    /// Called when an [acknowledgement of frames](FrameAcknowledgements) is received from Downstream.
+    fn incoming_acknowledged_frames(&mut self, ack: FrameAcknowledgements<C>) -> Result<(), SessionError>;
+
+    /// Called when a complete Frame has been finalized from segments received from Downstream.
+    fn frame_received(&mut self, id: FrameId) -> Result<(), SessionError>;
+
+    /// Called when a frame could not be completed from the segments received from Downstream.
+    fn frame_discarded(&mut self, id: FrameId) -> Result<(), SessionError>;
+
+    /// Called when a segment of a Frame was sent to the Downstream.
+    fn segment_sent(&mut self, segment: &Segment) -> Result<(), SessionError>;
+}
+
+/// Represents a stateless Session socket.
+/// Does nothing by default, only logs warnings and events for tracing.
+#[derive(Clone)]
+pub struct Stateless<const C: usize>(String);
+
+impl<const C: usize> Stateless<C> {
+    pub fn new<I: std::fmt::Display>(session_id: I) -> Self {
+        Self(session_id.to_string())
+    }
+}
+
+impl<const C: usize> SocketState<C> for Stateless<C> {
+    fn session_id(&self) -> &str {
+        &self.0
+    }
+
+    #[tracing::instrument(skip(self), fields(session_id = self.0))]
+    fn run(&mut self, _: SocketComponents<C>) -> Result<(), SessionError> {
+        tracing::debug!("stateless socket started");
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(self), fields(session_id = self.0))]
+    fn stop(&mut self) -> Result<(), SessionError> {
+        tracing::debug!("stateless socket stopped");
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(self, seg_id), fields(frame_id = seg_id.0, seq_idx = seg_id.1))]
+    fn incoming_segment(&mut self, seg_id: &SegmentId, seq_len: SeqNum) -> Result<(), SessionError> {
+        tracing::trace!("incoming segment");
```

Comment:
Don't repeat the tracing, just extend the macro with appropriate config if necessary.